querier_test.go

```diff
@@ -1691,6 +1692,118 @@ func BenchmarkQuerySeek(b *testing.B) {
 	}
 }
 
+func BenchmarkSetMatcher(b *testing.B) {
+	cases := []struct {
+		numBlocks                   int
+		numSeries                   int
+		numSamplesPerSeriesPerBlock int
+		setPattern                  string
+		regexPattern                string
+	}{
+		{
+			numBlocks:                   1,
+			numSeries:                   15,
+			numSamplesPerSeriesPerBlock: 10,
+			setPattern:                  "^(?:1|2|3)$",
+			regexPattern:                "1|2|3",
+		},
+		{
+			numBlocks:                   1,
+			numSeries:                   15,
+			numSamplesPerSeriesPerBlock: 10,
+			setPattern:                  "^(?:1|2|3|4|5|6|7|8|9|10)$",
+			regexPattern:                "1|2|3|4|5|6|7|8|9|10",
+		},
+		{
+			numBlocks:                   1,
+			numSeries:                   200,
+			numSamplesPerSeriesPerBlock: 10,
+			setPattern:                  "^(?:1|2|3)$",
+			regexPattern:                "1|2|3",
+		},
+		{
+			numBlocks:                   1,
+			numSeries:                   200,
+			numSamplesPerSeriesPerBlock: 10,
+			setPattern:                  "^(?:1|2|3|4|5|6|7|8|9|10)$",
+			regexPattern:                "1|2|3|4|5|6|7|8|9|10",
+		},
+	}
+
+	for _, c := range cases {
+		dir, err := ioutil.TempDir("", "bench_postings_for_matchers")
+		testutil.Ok(b, err)
+		defer func() {
+			testutil.Ok(b, os.RemoveAll(dir))
+		}()
+
+		var (
+			blocks          []*Block
+			prefilledLabels []map[string]string
+			generatedSeries []Series
+		)
+		for i := int64(0); i < int64(c.numBlocks); i++ {
+			mint := i * int64(c.numSamplesPerSeriesPerBlock)
+			maxt := mint + int64(c.numSamplesPerSeriesPerBlock) - 1
+			if len(prefilledLabels) == 0 {
+				generatedSeries = make([]Series, c.numSeries)
+				for i := 0; i < c.numSeries; i++ {
+					lbls := make(map[string]string, 10)
+					// The first label pair is {"test", "i"} which is for benchmarking set matcher.
+					lbls["test"] = strconv.Itoa(i)
```

Comment:
Can we repeat the same `i` for more than 1 series? If we are having a higher number of series now, we can maybe have this as `i%someNumber` (50-100 is what is in my mind).